module4/persistence/src/main/java/com/epam/esm/persistence/impl/CertificateDaoImpl.java

```diff
@@ -0,0 +1,51 @@
+package com.epam.esm.persistence.impl;
+
+import com.epam.esm.criteria_info.CertificateCriteriaInfo;
+import com.epam.esm.criteria_info.PageInfo;
+import com.epam.esm.entity.Certificate;
+import com.epam.esm.persistence.CertificateDao;
+import com.epam.esm.persistence.dataspecification.CertificateSpecification;
+import com.epam.esm.persistence.repository.CertificateRepository;
+import lombok.RequiredArgsConstructor;
+import org.springframework.data.domain.PageRequest;
+import org.springframework.data.domain.Sort;
+import org.springframework.stereotype.Repository;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+@RequiredArgsConstructor
+@Repository
+public class CertificateDaoImpl implements CertificateDao {
+
+    private final CertificateRepository certificateRepository;
+
+    @Override
+    public List<Certificate> findAll(CertificateCriteriaInfo criteriaInfo, PageInfo pageInfo, Sort sort) {
+        CertificateSpecification specification = new CertificateSpecification(criteriaInfo);
+        return certificateRepository.findAll(specification, PageRequest.of(pageInfo.getCurrentPage() - 1, pageInfo.getLimit(), sort))
+                .stream()
+                .collect(Collectors.toList());
```

Comment:
Why do you use stream-collect-toList? As far as I rememder, findAll returns List.

Reply:
if the method takes a Pageable type parameter, it returns Page<T>

Reply:
ah, exactly.
Why you don't want to use Page?

Reply:
beacause if findAll take type Pageable parametr it return `Page`

Reply:
use page